compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -1576,6 +1576,45 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         false
     }
 
+    /// When a failed try operator follows an `unwrap()` or `expect()`, suggest removing it
+    fn suggest_remove_unwrap(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diag<'_>,
+    ) -> bool {
+        if !matches!(obligation.cause.code(), ObligationCauseCode::QuestionMark) {
+            return false;
+        }
+
+        let Some(body) = self.tcx.hir().maybe_body_owned_by(obligation.cause.body_id) else {
+            return false;
+        };
+
+        let mut expr_finder = super::FindExprBySpan::new(obligation.cause.span, self.tcx);
+        expr_finder.visit_expr(body.value);
+
+        if let Some(expr) = expr_finder.result
+            && let Some(node) = self.tcx.hir().parent_iter(expr.hir_id).nth(2)
+            && let hir::Node::Expr(expr) = node.1
+            && let hir::ExprKind::Match(expr, _, _) = expr.kind
+            && let hir::ExprKind::Call(_, args) = expr.kind
+            && let Some(expr) = args.get(0)
+            && let hir::ExprKind::MethodCall(pathsegment, receiver, _, span) = expr.kind
+            && (pathsegment.ident.name == sym::unwrap || pathsegment.ident.name == sym::expect)
```

Comment:
This is not the best way to handle things. I would want these to be diagnostic items and the `DefId`s matched

Reply:
I meant the methods should be diagnostic items. We shouldn't use symbol matching in diagnostic code

Reply:
Note that in *some cases* the res for the method being accessed is `Err`. In those cases it becomes more reasonable *as a fallback*.

Reply:
^ This seems to be the case here; in fact, I don't know how to construct a test case where res is *not* `Err`:

```
kind: MethodCall(
    PathSegment {
        ident: expect#0,
        hir_id: HirId(DefId(0:3 ~ try_after_unwrap_issue_127345[6db5]::foo).4),
        res: Err,
        args: None,
        infer_args: true,
    },
```

Reply:
Also https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-items.html#naming-conventions says "Associated items should not get their own diagnostic items, but instead be accessed indirectly by the diagnostic item of the type they're originating from."

Reply:
Okay, makes sense.